pipeline/store/accessor.go

```diff
@@ -0,0 +1,177 @@
+		if err != nil {
+			return fmt.Errorf("failed to hydrate store %q: %w", store.Name, err)
+		}
+	}
+	return nil
+}
+
+func (s *Accessor) ResetStores() {
+	for _, store := range s.storeMap {
+		store.Reset()
+	}
+}
+
+// will attempt tp flush the stores  (or stores in filter if present) and report back the saved ranges
+func (r *Accessor) Flush(ctx context.Context, blockNum uint64, stopBlockNum uint64, storeFilter map[string]bool) (block.Ranges, error) {
+	var out block.Ranges
+	if err := r.flushLoop(blockNum, stopBlockNum, func() error {
+		blockRanges, err := r.flushStores(ctx, r.nextSaveBoundary, storeFilter)
+		if err != nil {
+			return fmt.Errorf("failed to flush stores: %w", err)
+		}
+		out = append(out, blockRanges...)
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	return out, nil
+}
+
+func (r *Accessor) flushLoop(blockNum uint64, stopBlockNum uint64, cb func() error) error {
+	for r.nextSaveBoundary <= blockNum {
+		r.logger.Debug("flushings stores", zap.Uint64("block_num", blockNum), zap.Uint64("boundary", r.nextSaveBoundary))
+
+		if err := cb(); err != nil {
+			return err
+		}
+
+		r.nextSaveBoundary = r.computeBoundaryBlock(r.nextSaveBoundary)
+		if isStopBlockReached(blockNum, stopBlockNum) {
+			break
+		}
+	}
+	return nil
+}
+
+func (r *Accessor) flushStores(ctx context.Context, blockNum uint64, storeFilter map[string]bool) (out block.Ranges, err error) {
+	for _, store := range r.storeMap {
+		// if the filter is set only flush these stores.
+		if len(storeFilter) > 0 && !storeFilter[store.Name] {
+			continue
+		}
+
+		savedBlockRange, err := r.flushStore(ctx, store, blockNum)
+		if err != nil {
+			return nil, fmt.Errorf("failed to flush store %q: %w", store.Name, err)
+		}
+		out = append(out, savedBlockRange)
+	}
+	return out, nil
+}
+
+func (r *Accessor) flushStore(ctx context.Context, store *state.Store, boundaryBlock uint64) (*block.Range, error) {
+	stateWriter, err := store.WriteState(ctx, boundaryBlock)
+	if err != nil {
+		return nil, fmt.Errorf("store writer %q: %w", store.Name, err)
+	}
+	if err := stateWriter.Write(); err != nil {
+		return nil, fmt.Errorf("writing store '%s' state: %w", store.Name, err)
+	}
+	r.logger.Info("store written", zap.String("store_name", store.Name), zap.Object("store", store))
+	return block.NewRange(store.StoreInitialBlock(), boundaryBlock), nil
+}
+
+func (s *Accessor) Iter(cb func(name string, store *state.Store) error) error {
+	for name, store := range s.storeMap {
+		if err := cb(name, store); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *Accessor) MarshalLogObject(enc zapcore.ObjectEncoder) error {
+	enc.AddInt("store_count", len(r.storeMap))
+	return nil
+}
+
+func (r *Accessor) computeBoundaryBlock(atBlockNum uint64) uint64 {
+	boundary := atBlockNum - atBlockNum%r.SaveInterval + r.SaveInterval
+	if r.maxBoundaryBlockNum > 0 && r.maxBoundaryBlockNum < boundary {
+		return r.maxBoundaryBlockNum
+	}
+	return boundary
+}
```

Comment:
I recall we needed the distinction between `initStoreSaveBoundary` and `bumpStoreSaveBoundary` and it was a subtle difference.

Now I only see `computeBoundaryBlock` and the tests that covered it were removed.